Dockerize Carnap & add GitHub Actions CI

lf- opened the pull request:
Hi! I have some more code for review :)

## This change set:

* Adds GitHub Actions continuous integration for Carnap and its dependencies that pushes outputs to Cachix. It also builds a Docker image. If tests are written and `doCheck` reenabled in the Nix files, this would also run them on every push as well. It would be fairly easy to copy paste the Carnap-Server building part of this Action to make it run on pull requests as well, we'd just have to make sure we aren't accidentally pushing stuff from PRs to Cachix (may be resolved automatically by how GitHub Actions handles secrets, I am not certain). We also now have cached Mac binaries for all the dependencies by virtue of building on Mac as well, making it an order of magnitude faster to get started developing on that platform.

* Moves the signing key from `(current working directory)/config/client-session-key.aes` to `$DATAROOT/client-session-key.aes` in order to try to put all the writable stuff in one place (still a WIP situation though since the book folder and the static folder still need to be writable due to caching of things in both as far as I can tell).

* Makes the example config fully generic: everything can now be specified in environment variables, which means we can directly ship it as the default config for Docker.

* Adds gitignore support to our packages. This makes sure there's no gitignored stuff such as temporary files from Carnap-Server's static folder getting into Nix builds and causing pointless rebuilds.

* Substantially reduces the size of the Carnap-Server package and its dependencies (it is now 3GB, it used to be 11GB 🤯) by building it statically and removing references that were causing unused runtime dependencies.

* According readme changes.

## Stuff that needs to be arranged:

* Slight breakage of existing Carnap installations since they need the session key moved to their data directory.

* Cachix signing key needs to be added to the repository settings. If you want to make your own Cachix other than `jade-carnap`, feel free! (more on this on Matrix)

lf- committed:
initial pass on runnable carnap-server package

lf- committed:
Gitignore our projects

In general, source filtering in Nix works well, but we have particular
files that should be ignored that are not handled by default. As such,
use a gitignore implementation for Nix to add it.

lf- committed:
Make it possible to turn on/off profiling

lf- committed:
initial try at docker

lf- committed:
Add CI

lf- committed:
delete superfluous command that broke the build lol

lf- committed:
add support for building without kvm

lf- committed:
fix docker build command

lf- committed:
Add container upload

lf- committed:
Add container upload

lf- committed:
make universal

lf- committed:
Build binaries for macOS

lf- committed:
stop pushing docker images to cachix

lf- committed:
massively reduce container size, we hope

lf- committed:
Attempt to further drastically reduce image size

lf- committed:
Rip out a bunch of closure size indiscriminantly**

TODO/NOTE: currently diagrams-builder is broken. This is not a
regression, it didn't work to begin with with nix-built Carnap-Server
(we need to provide the appropriate runtime dependencies which we don't
do yet).

lf- committed:
More closure size reduction

lf- committed:
Add an entrypoint to the docker container

lf- committed:
Finalize experimental docker support

gleachkr reviewed (commented):
Awesome! Looks exciting. I'll start reviewing tomorrow.

lf- committed:
Create the book cache folder, oops

lf- committed:
Set up CI for pull requests also

We don't push other peoples' artifacts to our cachix (probably already
not a problem but let's be as explicit as possible).

lf- committed:
just provide "latest" tag

gleachkr commented on server.nix:
```diff
@@ -33,27 +38,83 @@
+                  cp ${client-ghcjs.out}/bin/AllActions.jsexe/lib.js static/ghcjs/allactions/
+                  cp ${client-ghcjs.out}/bin/AllActions.jsexe/runmain.js static/ghcjs/allactions/
+                  echo ":: Adding a universal settings file"
+                  cp config/settings-example.yml config/settings.yml
```
Will this overwrite an existing settings file? Perhaps add `-n`?

gleachkr replied:
May also want to update `make run` appropriately.

lf- replied:
> Will this overwrite an existing settings file? Perhaps add `-n`?

Nope, it is running in a build directory with a copy of the source tree after gitignore is applied. It will override the settings.yml, yes, but it would also do that with `-n` due to the gitignore being applied to the source tree so any local copy would not be copied into the build. Regardless, this bit is only used for packaging and docker for which it is reasonable (for closure size/image size) to just copy the ghcjs stuff we use so we don't accidentally pull in node &etc for the executable outputs as we would if we symlinked.

`make run` is still ok for dev purposes.

gleachkr replied:
OK, thanks, I understand what's going on a bit better now.

gleachkr commented on README.md:
```diff
@@ -128,16 +128,55 @@ $ nix-shell -A ghcjsShell
+broken, which impacts some chapters of the Carnap textbook.
+
+Images are available via the GitHub container registry at
+`docker.pkg.github.com/gleachkr/carnap/carnap:GITREVISION`.
```
might want to change since the main repo has moved from gleachkr/carnap to carnap/carnap.

lf- replied:
Yup. That one snuck in after I wrote the readme. Changed.

lf- committed:
Update README to reflect the repo + cachix move

lf- committed:
switch CI to the official Carnap cachix
